File: c/scrabble-score/scrabble_score.c
#include "scrabble_score.h"

#include <ctype.h>
#include <string.h>

unsigned int letter_score(char letter);

// Calculates the score of a given word in the game of scrabble
unsigned int score(const char *word) {
  static const unsigned int initial_score = 0;

  if (!word) {
    return initial_score;
  }

  const size_t word_len = strlen(word);
  unsigned int total_score = initial_score;

  for (size_t i = 0; i < word_len; i++) {
    const unsigned int cscore = letter_score(word[i]);
    total_score += cscore;
  }

  return total_score;
}

#define ALPHABET_LEN 26
// Calculates the index of a capital letter in the alphabet.
#define indexof(letter) letter - 'A'

// Stores the score of each letter
int scores[ALPHABET_LEN] = {
    [indexof('A')] = 1,  [indexof('E')] = 1, [indexof('I')] = 1,
    [indexof('O')] = 1,  [indexof('U')] = 1, [indexof('L')] = 1,
    [indexof('N')] = 1,  [indexof('R')] = 1, [indexof('S')] = 1,
    [indexof('T')] = 1,  [indexof('D')] = 2, [indexof('G')] = 2,
    [indexof('B')] = 3,  [indexof('C')] = 3, [indexof('M')] = 3,
    [indexof('P')] = 3,  [indexof('F')] = 4, [indexof('H')] = 4,
    [indexof('V')] = 4,  [indexof('W')] = 4, [indexof('Y')] = 4,
    [indexof('K')] = 5,  [indexof('J')] = 8, [indexof('X')] = 8,
    [indexof('Q')] = 10, [indexof('Z')] = 10};

// Calculates and returns the score of a letter in the scrabble game. If the
// character is not a letter, it returns 0.
unsigned int letter_score(char letter) {
  if (isupper(letter)) {
    return scores[indexof(letter)];
  }
  if (islower(letter)) {
    return scores[indexof(toupper(letter))];
  }
  static const unsigned int no_score = 0;
  return no_score;
}

File: php/scrabble-score/ScrabbleScore.php
<?php

declare(strict_types=1);

/**
 * Calculates the Scrabble score for a given word.
 *
 * This function sums the individual scores of each letter in the word.
 * It normalizes the input to uppercase before calculating the score.
 * Characters not defined in Scrabble scoring (e.g., spaces, punctuation) contribute 0 points.
 *
 * @param string $word The word for which to calculate the Scrabble score.
 * @return int The total Scrabble score for the word.
 *
 * @example
 * score('a');              // Returns 1
 * score('f');              // Returns 4
 * score('at');             // Returns 2
 * score('zoo');            // Returns 12
 * score('street');         // Returns 6
 * score('quirky');         // Returns 22
 * score('OxyphenButazone'); // Returns 41
 * score('');               // Returns 0
 * score('abcdefghijklmnopqrstuvwxyz'); // Returns 87
 */
function score(string $word): int
{
    $score = 0;

    foreach (normalizeChars($word) as $letter) {
        $score += letterScore($letter);
    }

    return $score;
}

/**
 * Returns the Scrabble score for a single letter.
 *
 * The scoring follows standard English Scrabble rules.
 *
 * @param string $letter The single letter (expected to be uppercase) to score.
 * @return int The Scrabble score for the letter. Returns 0 for invalid/unscored characters.
 */
function letterScore(string $letter): int
{
    return match ($letter) {
        'A', 'E', 'I', 'O', 'U', 'L', 'N', 'R', 'S', 'T' => 1,
        'D', 'G' => 2,
        'B', 'C', 'M', 'P' => 3,
        'F', 'H', 'V', 'W', 'Y' => 4,
        'K' => 5,
        'J', 'X' => 8,
        'Q', 'Z' => 10,
        default => 0
    };
}

/**
 * Normalizes a given text string by converting it to uppercase and splitting it into an array of individual characters.
 *
 * This is a helper function to prepare the word for letter-by-letter scoring.
 *
 * @param string $text The input string to normalize.
 * @return array<string> An array of uppercase characters.
 */
function normalizeChars(string $text): array
{
    return str_split(strtoupper($text));
}